README
###########################################################################
### Frogesport                                                          ###
###########################################################################
###
#
#    Copyright 2011-2014 Zmegolaz <zmegolaz@kaizoku.se>
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

REQUIREMENTS
Eggdrop version 1.8 or later, this is where the procedure putnow is introduced, and it's used to send messages past the flood fitlered queues.
A MySQL database.
TCL package mysqltcl.

COMMANDS
Command - Description - Requirements
!answer - Get the answer for the last question - OP
!clear - Clear all pending actions, should never have to be used - Admin
!clearqueue - Clear the queue of questions to ask next - Depends on configuration
!compare - Compare any number of users to eachother, ordered by season points - Anyone
!comparetot - Compare any number of users to eachother, ordered by total points - Anyone
!continue - If the bot has been paused, start it again - Admin
!hof - Hall of fame, the 10 users with the highest total points - Anyone
!hofclan - Hall of fame, the 10 clans with the highest total points - Anyone
!pause - Pause the bot, without clearing the questions IDs - Admin
!punish - Take points from a user - Admin
!queue [number] <string/id> - Add one or more question matching the search string to the queue of questions to ask next - Depends on configuration
!reward - Give points to a user - Admin
!spell <spell> [nick|id] - Cast a spell - Anyone of level 5 or higher
!startmess <interval> <message> - Start sending a message every <interval> minutes - Admin
!startquiz - Start the bot - Admin
!stats [nick] - Check the stats for someone, if [nick] is omitted, you will get your own stats - Anyone
!stopmess - Stop sending a message every x minutes - Admin
!stopquiz - Stop the bot - Admin
!recommend - Get help about how to recommend questions - Anyone
!report <id> <comment> - Report a question - Anyone
!time - Shows time left until the user can switch channel - Anyone
!top10 - Top 10 users of this season - Anyone
!top10clan - Top 10 clans of this season - Anyone
!version - Show the bot version - Anyone

[PM]checkq <id> [perm] - Check any question and it's answers - Admin
[PM]checku <nick> - Check a user, an alias to !stats - Admin
[PM]clearqueue - Clear the queue of questions to ask next - Depends on configuration
[PM]addq <category>|<question>|<answer>\[|answer\]... - Add a question - Admin
[PM]delq <id> [perm] - Delete a question - Admin
[PM]modq <id> [perm] <category|question|answer> <value> - Modify a part of a question - Admin
[PM]modu <nick|season|total|time|inarow|mana|class|last|customclass|transfer|help> <nick|subcommand> [new value] - Modify any part of a user - Admin
[PM]newseason [help] - Start a new season - Admin
[PM]updateclasses - Correct all users' classes - Admin
[PM]queue [number] <string/id> - Add one or more question matching the search string to the queue of questions to ask next - Depends on configuration
[PM]recommend <category>|<question>|<answer>[|<answer>] - Recommend a question - Anyone
[PM]clan create <clan name>|join <clan name>|leave|role <nick> <member|admin> <yes|no>|change name <new name>|members

TODO, no internal order:
Make sure all all MySQL indexes are as they should be.
Newseason should save the old table as users_unixtime.
Newseason should remove inactive users with low (configurable) total points.
Newseason should reset the mana too.
Send some kind of message if someone tries to compare a user to the same user.
Move all text to a separate file for easier translations.
Translate everything to English.
Documentation.
Logg all user/question modifications.
Happy hour.
Config sanity check.
Separate the help, stats and similar from the question asking.
Best time per question.
Number of times a question is asked, and number of times it was answered correctly.
Different points for different questions.


CHANGELOG
Version 1.8 - 2014-11-20
	NOTE: This version uses a new database structure, make sure to update! An SQL file to use when updating is provided. Apply database updates before rehashing if update is done on a live bot.
	Features
 * Users can now join clans and compete as a team. Channel commands !top10clan, !top10clanavg, !hofclan and !hofclanavg, as well as PM command "clan" has been added.
	Changes
 * The question ID is now underlined.
	Bug fixes
 * A lot of arguments failed if they had unmatched quotes or brackets in them.
 * Nicks with [] in them wasn't handled correctly in some cases
 * Fixed a dangling pointer in rare cases when the database disappeared.
Version 1.7 - 2014-06-03
	Features
 * It is now possible to manually tell the bot to enqueue questions to ask next, by ID or by searching for a string.
Version 1.6.2 - 2014-04-05
	Changes
 * !startmess now takes two arguments, interval and message, instead of them having to be configured in hte config file.
Version 1.6.1 - 2014-03-31
	Features
 * You can now use !topkpm and !topfast to see the top lists of the respective columns.
Version 1.6 - 2014-03-17
	NOTE: This version uses a new database structure, make sure to update! An SQL file to use when updating is provided. Apply database updates before rehashing if update is done on a live bot.
	Features
 * KPM is now displayed when an answer was correct, and top KPM is saved in the database.
Version 1.5 - 2014-01-03
	NOTE: This version uses a new database structure, make sure to update! An SQL file to use when updating is provided. Apply database updates before rehashing if update is done on a live bot.
	Changes
 * It is now supported to run multiple bots simultaneously, using the same database.
 * Whitespace before and after an answer is ignored.
	Features
 * It is now possible to send multiple levels of clues at different intervals.
 * A configuration option to save all correct answers is added. For now we're not doing anything with them, but it opens the possibility to provide per user statistics on a web site.
 * !reward and !punish now has an optional argument with the amount of points to give/take.
 * An answer can be marked as preferred, which means it is preferred when creating clues.
Version 1.4 - 2013-09-24
	NOTE: This version uses a new database structure, make sure to update! An SQL file to use when updating is provided. Apply database updates before rehashing if update is done on a live bot.
	Changes
 * modu transfer now moves customclass too, if the source user has one.
	Bug fixes
 * The recommendation system didn't insert the recommendations correctly. (oups)
	Features
 * !suggest is now an alias for !recommend.
 * Suggestions (recommendations) and questions now has a column for source.
 * The nick of the user adding a question is now saved.
 * Suggestions now have a lastshow feature like recommendations, so that multiple admins can work simultaneously.
Version 1.3.2 - 2013-09-07
	Bug fixes
 * The spell prevanswer required a new question to have been asked, even if the previos question was finished.
Version 1.3.1 - 2013-04-14
	Changes
 * Changed MySQL collation to utf8_unicode_ci.
 * Changed MySQL connection charset to iso8859-1, this seemed to be the only one which could handle special characters from clients (even utf8 ones).
	Bug fixes
 * !compare failed when there was a trailing space.
Version 1.3 - 2013-03-17
	Features
 * The bot now saves the time each report was shown to an admin and prefers to show others the next time. This makes it possible for multiple admins to correct reports at the same time, without disturbing eachother. Note that the database structure has been changed!
	Bug fixes
 * The last change regardin MySQL connections caused the bot to fail to start.
Version 1.2.3.2 - 2013-03-16
	Bug fixes
 * Changed the database reconnect thingie on rehash, if the SQL server went away it didn't reconnect as it should.
Version 1.2.3.1 - 2013-02-23
	Bug fixes
 * When deleting a question, the answers and reports related to it were not deleted.
Version 1.2.3 - 2012-11-17
	Changes
 * Put everything in a namespace to better separate variable and procedure names from other code.
	Bug fixes
 * Sometimes (after a couple of days) the connection to the database was dropped, disabling everything but the questions themselves.
 * The close_behind time check didn't set the correct value.
 * When someone gaind a rank the bot stopped.
 * !top10 and !hof didn't work.
Version 1.2.2.1 - 2012-11-16
	Bug fixes
 * Fixed a bug where checkq output the answers in an incorrect way.
 * Moved the bind to read the answers.
Version 1.2.2 - 2012-11-16
	Features
 * The close answers feature now applies to questions where nobody answered correctly too.
	Changes
 * A lot of SQL optimizations to speed up everything and use less memory.
 * Changed charset in the database tables.
Version 1.2.1 - 2012-11-15
	Features
 * Added a "close answers" to show people how close they were to winning the question.
	Changes
 * Rewrote the randomization function to greatly simplify it and the database structure.
 * Included the SQL table structure (yeah, should have done this earlier...)
	Bug fixes
 * Changed to correct syntax for modq in help.
Version 1.2.0 - 2012-11-11
	Changed name to Frogesport. "Frogesport" is pronounced similar to the swedish word "frågesport" which means "quiz", and it's a frog themed e-sport. Funny, right?
Version 1.1.27 - 2012-02-18
	Features
 * !pause was introduced.
 * !startmess and !stopmess can be used to start/stop sending a message every half hour.
Version 1.1.26 - 2011-12-02
	Bug fixes
 * The spell prevanswer could be used to get the answer to the current question too.
Version 1.1.25 - 2011-10-07
	Features
 * The PM command modu transfer is now added, which transfers all stats from one user to another.
Version 1.1.24 - 2011-10-06
	Changes
 * The start and stop commands are now !startquiz and !stopquiz

frogesport-config-EXAMPLE.tcl
###########################################################################
### Frogesport                                                          ###
###########################################################################
###
#
#    Copyright 2011-2014 Zmegolaz <zmegolaz@kaizoku.se>
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
###

############################################################
###                    Configuration                     ###
############################################################

# Edit these to match your MySQL settings
variable mysql_user "frogesport"
variable mysql_pass "secret"
variable mysql_host "localhost"
variable mysql_dbname "frogesport"

# Channel the bot is used in, this has to be correct to be able to start the quiz
variable running_chan "#channel"

# Bot ID. Used when multiple bots are run with the same database, make sure each bot has it's own unique ID. Any number between 0 and 99999 can be used.
variable i_bot_id "1"

# Should we save all correct answers?
variable save_answers "yes"

# Set authentication method
# 1: Check if the user is op in the running channel and if the user is class 0 in the database
# 2: Check if the user is op in the running channel and is in the configured list of admins
# 3: Check if the user is op in the admin channel
variable auth_method "2"

# Bot admins, separated by space. Only needed if auth_method is set to 2
variable admins "your-nick"
# Channel with the bot admins. Only needed if auth_method is set to 3
variable admin_chan "#channel-admin"

# Time between questions, in seconds
variable s_question_time "10"
# Time between each clue, in seconds
variable s_clue_time "10"
# Time users have to answer the question, in seconds
variable s_time_answer "30"
# The bot may tell people how long time behind the winner they were, for how long should it collect nicks?
# Enter in seconds, and has to be lower than s_question_time. Set to 0 to disable this feature.
variable s_close_behind "3"
# Time the user has to wait between answering a question in two different channels, in seconds.
variable s_channel_switch_time "240"

# How many percent of each word in the answer the clue should display, number of letters are rounded up.
# You can have multiple percentages here (in the same quotation marks) to give multiple clues.
variable clue_percent "30 60"

# Set this to 1 if the bot should give the answers if no user answered correctly
variable give_answer "0"

# At which level should users be allowed to cast spells?
variable spell_level "5"
# Cost of the spells
variable cost_steal "50"
variable cost_give "50"
variable cost_answer "50"
variable cost_setvoice "10"
variable cost_prevanswer "20"
# How many points should be stolen or given?
variable steal_points "5"
variable give_points "5"
# How many points should !punish and !reward take/give?
variable rewpun_points "5"

# How many reports nad recommended questions should be shown in PM?
# Be careful, to high and the bot'll be too busy pasting reports and won't answer other commands
# A web interface to check all is preferred
variable reports_show "7"
variable recommend_show "7"

# Who should be able to queue specific questions?
# 0: Nobody.
# 1: Admins (see auth_method above.)
# 2: OPs and admins.
# 3: Voiced users, OPs and admins.
# 4: Anyone.
variable queueques_who 0

# How many questions should be allowed to be in the queue?
variable queueques_num 30
# Where should we search for the keywords?
# 1: Category column.
# 2: Question column.
# 3: Both category and question columns.
variable queueques_searchwhere 1

# How many seconds has to pass between each ping command?
variable s_pinginterval "10"

# Text color:
variable color_text "00"
# Text background color.
variable color_background "01"
# Color of users nicks.
variable color_nick "09"
# Color of classes.
variable color_class "08"
# Color of answers.
variable color_answer "04"
# Color of numbers in stats.
variable color_statsnumber "04"

# Colors:
#00 white
#01 black
#02 blue (navy)
#03 green
#04 red
#05 brown (maroon)
#06 purple
#07 orange (olive)
#08 yellow
#09 light green (lime)
#10 teal (a green/blue cyan)
#11 light cyan (cyan) (aqua)
#12 light blue (royal)
#13 pink (light purple) (fuchsia)
#14 grey
#15 light grey (silver)

